docs/goodyara.md

```diff
@@ -0,0 +1,156 @@
+# How to write good Yara rules
+
+In the ideal world, every Yara rule would be supported equally well by mquery.
+Unfortunately, this is not the case. Some rules work great, but some are pretty slow.
+
+In this document, we will give some hints on how to make your rules more
+Mquery-friendly. It assumes you're already familiar with the Yara language.
+
+We will only cover the basic rules here. To learn more read [this](./yara.md)
+document with more details.
+
+## The basics
+
+Simplifying a lot, mquery works by first generating a list of "likely hits", and
+later running real yara on them. Prefiltering works by looking at all input
+strings, splitting them into short fragments (called n-grams), and looking
+them up separately. For example, this rule:
+
+```yara
+rule mquery_example
+{
+    strings:
+        $first = {11 22 33 44 55}
+    condition:
+        all of them
+}
+```
+
+Is looking for a sequence o bytes {11 22 33 44 55}. Internally, mquery will
+look for:
+
+- files containing a trigram {11 22 33}
+- files containing a trigram {22 33 44}
+- files containing a trigram {33 44 55}
+
+And AND the resulting sets together. False positives are later sorted out by
+running a normal Yara matching on candidate files.
+
+Not only 3grams are used. For plaintext strings and Unicode strings (a-zA-Z0-9)
+4-grams (n-grams of length 4, like "firs" and "irst") are used if enabled.
+
+Every condition that can't be checked using just n-grams will be ignored,
+- and this means that mquery will stay on the safe side and assume every file satisfies it.
+
+## In practice
+
+Ok, what does it mean in practice? It's easier to understand by looking at
+examples where mquery **doesn't** work well:
+
+### bad case 1
+
+```yara
+rule MAL_RedLeaves_Apr18_1 {
+   meta:
+      description = "Detects RedLeaves malware"
+      author = "Florian Roth"
+      reference = "https://www.accenture.com/t20180423T055005Z__w__/se-en/_acnmedia/PDF-76/Accenture-Hogfish-Threat-Analysis.pdf"
+      date = "2018-05-01"
+      hash1 = "f6449e255bc1a9d4a02391be35d0dd37def19b7e20cfcc274427a0b39cb21b7b"
+      hash2 = "db7c1534dede15be08e651784d3a5d2ae41963d192b0f8776701b4b72240c38d"
+      hash3 = "d956e2ff1b22ccee2c5d9819128103d4c31ecefde3ce463a6dea19ecaaf418a1"
+   condition:
+      uint16(0) == 0x5a4d and filesize < 1000KB and (
+         pe.imphash() == "7a861cd9c495e1d950a43cb708a22985" or
+         pe.imphash() == "566a7a4ef613a797389b570f8b4f79df"
+      )
+}
+```
+
+There are three things checked here:
+
+- First two bytes are 0x4d and 0x5a
+- File is smaller than 1MB
+- PE imphash is either "7a861cd9c495e1d950a43cb708a22985" or "566a7a4ef613a797389b570f8b4f79df".
+
+You probably see where this is going - all of these conditions are ignored.
+Mquery doesn't care about 2-byte fragments (or their location in the file), or file
```

Comment:
```suggestion
Mquery doesn't care about 2-byte fragments (or their location in the file), or file
```

not sure if this shouldn't be lowercase `mquery` as well?

Reply:
It's a start of a sentence - so gramatically speaking probably should be capitalised? I usually try to avoid starting a sentence with proper names to avoid this kind of dillema, but if it happens like in this case it probably should start with a capital letter?